text/0034-ternary-hash.md

```diff
@@ -0,0 +1,415 @@
+will not be cleared unless `reset` is called.
+
+### Design of `CurlP`
+
+`CurlP` is designed as a hash function that acts on a `T1B1` binary-encoded ternary buffer, with a hash length of `243`
+trits and an inner state of `729` trits, which we take as constants:
+
+```rust
+const HASH_LENGTH: usize = 243;
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+```
+
+In addition, a lookup table is used as part of the absorption step.
+2 is used to pad the table since there are only 9 combinations should be taken:
+
+```rust
+const TRUTH_TABLE: [i8; 11] = [1, 0, -1, 2, 1, -1, 0, 2, -1, 1, 0];
+```
+
+The way `CurlP` is defined, it can not actually fail, because the input or outputs can be of arbitrary size.
+Thus, the associated type `Error = !`, or `Error = Infallible` as `!` is not yet stabilized in Rust as of version
+`1.43`.
+
+This `enum` defines the number of rounds of hashing to apply before a hash is squeezed.
+```rust
+#[derive(Copy, Clone)]
+pub enum CurlPRounds {
+    Rounds27 = 27,
+    Rounds81 = 81,
+}
+```
+
+Type definition:
+
+```rust
+struct CurlP {
+    /// The number of rounds of hashing to apply before a hash is squeezed.
+    rounds: CurlPRounds,
+
+    /// The internal state.
+    state: TritBuf,
+
+    /// Workspace for performing transformations
+    work_state: TritBuf,
+}
+```
+
+In addition, there are two wrapper types for the very common `CurlP` variants with `27` and `81` rounds:
+
+```rust
+struct CurlP27(CurlP);
+
+impl CurlP27 {
+    pub fn new() -> Self {
+        Self(CurlP::new(CurlPRounds::Rounds27))
+    }
+}
+
+struct CurlP81(CurlP);
+
+impl CurlP81 {
+    pub fn new() -> Self {
+        Self(CurlP::new(CurlPRounds::Rounds81))
+    }
+}
+```
+
+### Design of `Kerl`
+
+The actual cryptographic hash function underlying `Kerl` is `keccak-384`. The actual task here is to transform an input
+of 243 (balanced) trits to 384 bits in a correct and performant way. This is done by interpreting the 243 trits as a
+signed integer `I` and converting it to a binary basis. In other words, the ternary coded integer is expressed as the
+series:
+
+```
+I = t_0 * 3^0 + t_1 * 3^1 + ... + t_241 * 3^241 + t_242 * 3^242,
+```
+
+where `t_i` is the trit at the `i`-th position in the ternary input array. The challenge is then to convert this integer
+to base `2`, i.e. find a a series such that:
+
+```
+I = b_0 * 2^0 + b_1 * 2^1 + ... + b_382 * 2^382 + b_383 * 2^383,
+```
+
+with `b_i` the bit at the `i-th` position.
+
+Assuming there exists an implementation of `keccak`, the main work in implementing `Kerl` is writing an efficient
+converter betWe en the ternary array interpreted as an integer, and its binary representation. For the binary
+representation, one can either use an existing big integer library, or write one from scratch with only a subset of
+required methods to make the conversion work.
+
+#### Important implementation details
+
```

Comment:
I would describe the whole algo in a bullet list before diving into the details (if that's even necessary):
Input: `I_bt` [BTrit; 243] **b**alanced **t**ernary trit array
* Shift `I_bt` into **u**nbalanced **t**ernary space to receive `I_ut` by adding the constant `H` [+1; 243] index-wise ,
* Calculate the **d**ecimal big-integer `I'_d` from `I_ut` by setting those as coefficients in the number series evaluation for base=3
* Determine the binary representation `bin(I'_d)` of `I'_d`,
* Calculate `bin(I_d) = bin(I'_d) - bin(H)`
* Apply `keccak384(bin(I_d))` ??

Is this the correct process? I wanted to better understand it myself, hence this comment.